google/ggrc-core
# GGRC-41 Add unsubscribe link to daily digest emails

This PR adds an "unsubscribe" link to the bottom of daily digest emails. By clicking the link, users can unsubscribe themselves from receiving daily digest emails.

Things left to do:
- [x] Clean up the code
- [x] Add integration tests for the new API endpoint
- [ ] Fix the "login as user" problem in integration tests
- [x] If a user is not logged in when clicking the link, he should be redirected to the login page first.
  Might already work? Have some trouble reproducing it locally)
- [x] If a user does not have an entry in the `notification_configs` DB table, unsubscribing does not work (because by default, users _are_ subscribed unless there is an entry in the aforementioned table that explicitly override that)
---

Questions:
- ~~Where should the `unsubscribe_from_notifications` be best placed in? Or is having it in `notifications.common` just fine?~~
- ~~How should the API endpoint look like, is having it under `_notifications/` fine?~~
- ~~Do we have any common error handlers for "not authorized", "internal server error", etc. scenarios? Right now those response messages and error codes are manually set.~~
- ~~Simply invoking the API with an email other than user@example.com results in 9 DB queries, even if the handler method body is completely commented out. Is this normal or could e.g. the API path definition be declared in a better way?
(mind that login must be required)~~
- ~~Should we log any attempts to unsubscribe a user other than the currently logged in user himself?~~

-----------------------

**Steps to verify:**
- Make sure a user has at least one notification pending, e.g. assign him/her to an Assessment and modify the latter
- Visit the daily digest page `_notifications/show_daily_digest`
- Find that user's name/email on the page and click the unsubscribe link at the bottom of the email.

**Expected result:**
A user should get unsubscribed. One can verify this by visiting the main application page and opening the user settings menu (top right corner). The _"Daily email digest"_ checkbox should be unchecked. Also, if refreshing the email notifications page, that user's email should not be there anymore.

**Additional notes:**
- When unsubscribing, users must be logged in in the application, and they can only unsubscribe themselves (their own email).
